WooCommerce/Classes/Authentication/Jetpack Setup/Native Jetpack Setup/JetpackSetupViewModel.swift

```diff
@@ -377,6 +332,163 @@ private extension JetpackSetupViewModel {
+            return handleSiteRegisterResult(isRegistered: isRegistered, blogID: data.blogID)
+        }
+
+        if isPluginActivated {
+            /// Skips plugin check if plugin has just got activated.
+            /// `isRegistered` is unavailable due to outdated Jetpack. Proceed with web flow.
+            startConnectionWithWebView()
+        } else {
+            /// Fetch plugin details to check
+            stores.dispatch(JetpackConnectionAction.retrieveJetpackPluginDetails { [weak self] result in
+                guard let self else { return }
+                switch result {
+                case .success:
+                    /// Plugin is available but`isRegistered` is unavailable due to outdated version.
+                    /// Proceed with web flow.
+                    startConnectionWithWebView()
+                case .failure(let error):
+                    if error.errorCode == 404 {
+                        /// For Jetpack-connected sites, if `isRegistered` is not returned,
+                        /// check for `connectionOwner` instead.
+                        handleSiteRegisterResult(isRegistered: data.connectionOwner != nil, blogID: data.blogID)
+                    } else {
+                        didFailJetpackConnection(with: error)
+                    }
+                }
+            })
+        }
+    }
+
+    func handleSiteRegisterResult(isRegistered: Bool, blogID: Int64?) {
+        if let blogID, isRegistered {
+            provisionSiteConnection(blogID: blogID)
+        } else {
+            registerSiteConnection()
+        }
+    }
+
+    func registerSiteConnection() {
+        stores.dispatch(JetpackConnectionAction.registerSite(completion: { [weak self] result in
+            guard let self else { return }
+            switch result {
+            case .success(let blogID):
+                provisionSiteConnection(blogID: blogID)
+            case .failure(let error):
+                didFailJetpackConnection(with: error)
+            }
+        }))
+    }
+
+    func provisionSiteConnection(blogID: Int64) {
+        currentSetupStep = .connection
+        trackSetupAfterLogin()
+        stores.dispatch(JetpackConnectionAction.provisionConnection(completion: { [weak self] result in
+            guard let self else { return }
+            switch result {
+            case .success(let response):
+                finalizeSiteConnection(blogID: blogID, provisionResponse: response)
+            case .failure(let error):
+                didFailJetpackConnection(with: error)
+            }
+        }))
+    }
+
+    func finalizeSiteConnection(blogID: Int64, provisionResponse: JetpackConnectionProvisionResponse) {
+        guard let wpcomCredentials, case .wpcom = wpcomCredentials else {
+            /// WPCom credentials are necessary to finalize connection through API
+            /// If this is unavailable, fall back to the web flow.
+            return startConnectionWithWebView()
+        }
```

Comment:
@itsmeichigo just a question about this part and the comment there, is there a case where we might reach this screen without WPCom credentials?

On Android, we don't have any check like this, and I looked into the iOS code, and it seems to me the only place where we pass an Optional value for the credentials is [here](https://github.com/woocommerce/woocommerce-ios/blob/d051c952adb9981e4b57dd0ff2b751ac3c53fad1/WooCommerce/Classes/Authentication/Jetpack%20Setup/LoginJetpackSetupCoordinator.swift#L50), where it seems the credentials would be necessarly present.

WDYT about this? If I'm right in my above remark, should we try to make the field in the ViewModel not optional to avoid any confusing because of this above check and comment?

Reply:
Thanks for the suggestion @hichamboushaba. I added the change in #16001.